nova/src/gadgets/cyclefold/hypernova/primary.rs

```diff
@@ -0,0 +1,561 @@
+use std::{borrow::Borrow, marker::PhantomData};
+
+use ark_crypto_primitives::sponge::constraints::{AbsorbGadget, SpongeWithGadget};
+use ark_ec::short_weierstrass::{Projective, SWCurveConfig};
+use ark_ff::{Field, PrimeField};
+use ark_r1cs_std::{
+    alloc::{AllocVar, AllocationMode},
+    boolean::Boolean,
+    fields::{fp::FpVar, FieldVar},
+    select::CondSelectGadget,
+    uint8::UInt8,
+    R1CSVar,
+};
+use ark_relations::r1cs::{ConstraintSystemRef, Namespace, SynthesisError};
+use ark_spartan::polycommitments::{PolyCommitmentScheme, PolyCommitmentTrait};
+use ark_std::fmt::Debug;
+
+use super::NonNativeAffineVar;
+use crate::folding::hypernova::cyclefold::{CCSInstance, HNProof, LCCSInstance};
+use crate::folding::hypernova::ml_sumcheck::PolynomialInfo;
+
+#[must_use]
+#[derive(Debug)]
+pub struct CCSInstanceVar<G1, C1>
+where
+    G1: SWCurveConfig,
+    G1::BaseField: PrimeField,
+{
+    /// Commitment to witness.
+    pub commitment_W: NonNativeAffineVar<G1>,
+    /// Public input of non-linearized instance.
+    pub X: Vec<FpVar<G1::ScalarField>>,
+
+    _commitment_scheme: PhantomData<C1>,
+}
+
+impl<G1, C1> Clone for CCSInstanceVar<G1, C1>
+where
+    G1: SWCurveConfig,
+    G1::BaseField: PrimeField,
+{
+    fn clone(&self) -> Self {
+        Self {
+            commitment_W: self.commitment_W.clone(),
+            X: self.X.clone(),
+            _commitment_scheme: self._commitment_scheme,
+        }
+    }
+}
+
+#[derive(Debug, Clone)]
+pub struct CCSInstanceFromR1CSVar<G1, C1>(CCSInstanceVar<G1, C1>)
```

Comment:
I wanted to make the implementation explicit about the necessary source of the constraints, as well as to interface with `ark-r1cs-std::R1CSVar` in a clear way, and this seemed like the most minimal way to do so. Happy to consider other approaches though.